pipenv/project.py

```diff
@@ -316,105 +317,48 @@ def working_set(self):
-            packages = [pkg for pkg in workingset if self.dist_is_in_project(pkg)]
-        else:
-            packages = [pkg for pkg in packages]
-        return packages
+    @property
+    def installed_packages(self):
+        return self.environment.get_installed_packages()
 
-    def get_package_info(self):
-        from .utils import prepare_pip_source_args
-        from .vendor.pip_shims import Command, cmdoptions, index_group, PackageFinder
-        index_urls = [source.get("url") for source in self.sources]
+    @property
+    def installed_package_names(self):
+        return get_canonical_names([pkg.key for pkg in self.installed_packages])
 
-        class PipCommand(Command):
-            name = "PipCommand"
+    @property
+    def lockfile_package_names(self):
+        dev_keys = get_canonical_names(self.lockfile_content["develop"].keys())
+        default_keys = get_canonical_names(self.lockfile_content["default"].keys())
+        return {
+            "dev": dev_keys,
+            "default": default_keys,
+            "combined": dev_keys | default_keys
+        }
 
-        dependency_links = []
-        packages = self.get_installed_packages()
-        # This code is borrowed from pip's current implementation
-        for dist in packages:
-            if dist.has_metadata('dependency_links.txt'):
-                dependency_links.extend(dist.get_metadata_lines('dependency_links.txt'))
+    @property
+    def pipfile_package_names(self):
+        dev_keys = get_canonical_names(self.dev_packages.keys())
+        default_keys = get_canonical_names(self.packages.keys())
+        return {
+            "dev": dev_keys,
+            "default": default_keys,
+            "combined": dev_keys | default_keys
+        }
 
-        pip_command = PipCommand()
-        index_opts = cmdoptions.make_option_group(
-            index_group, pip_command.parser
-        )
-        cmd_opts = pip_command.cmd_opts
-        pip_command.parser.insert_option_group(0, index_opts)
-        pip_command.parser.insert_option_group(0, cmd_opts)
-        pip_args = prepare_pip_source_args(self.sources, [])
-        pip_options, _ = pip_command.parser.parse_args(pip_args)
-        pip_options.cache_dir = PIPENV_CACHE_DIR
-        pip_options.pre = self.settings.get("pre", False)
-        with pip_command._build_session(pip_options) as session:
-            finder = PackageFinder(
-                find_links=pip_options.find_links,
-                index_urls=index_urls, allow_all_prereleases=pip_options.pre,
-                trusted_hosts=pip_options.trusted_hosts,
-                process_dependency_links=pip_options.process_dependency_links,
-                session=session
+    @property
+    def environment(self):
+        if not self._environment:
+            prefix = self.get_location_for_virtualenv()
+            is_venv = prefix == sys.prefix
+            sources = self.sources.copy() if self.sources else [DEFAULT_SOURCE,]
```

Comment:
```suggestion
            sources = self.sources[:] if self.sources else [DEFAULT_SOURCE,]
```

Is ```self.sources``` always a list? List does not have ```copy``` method.